x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/RemoteAccessAuthentication.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.transport;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentFactory;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xcontent.XContentParserConfiguration;
+import org.elasticsearch.xcontent.XContentType;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptorsIntersection;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+
+public record RemoteAccessAuthentication(Authentication authentication, Collection<BytesReference> roleDescriptorsBytesIntersection) {
+
+    private static final String REMOTE_ACCESS_AUTHENTICATION_HEADER = "_remote_access_authentication";
+
+    public static void writeToContext(
+        final ThreadContext ctx,
+        final Authentication authentication,
+        final RoleDescriptorsIntersection roleDescriptorsIntersection
+    ) throws IOException {
+        ctx.putHeader(REMOTE_ACCESS_AUTHENTICATION_HEADER, encode(authentication, roleDescriptorsIntersection));
+    }
+
+    static String encode(final Authentication authentication, final RoleDescriptorsIntersection roleDescriptorsIntersection)
+        throws IOException {
+        final BytesStreamOutput out = new BytesStreamOutput();
+        out.setVersion(authentication.getEffectiveSubject().getVersion());
+        Version.writeVersion(authentication.getEffectiveSubject().getVersion(), out);
+        authentication.writeTo(out);
+        out.writeVInt(roleDescriptorsIntersection.roleDescriptorsList().size());
+        for (var roleDescriptors : roleDescriptorsIntersection.roleDescriptorsList()) {
+            final XContentBuilder builder = XContentFactory.jsonBuilder();
```

Comment:
This and `parseRoleDescriptorBytes` can be cleaned up -- just providing an overall idea of how this will work here.